### ISSUE TITLE
Picture/Visualizer Architecture

### ISSUE THREAD

**Issue by Agizin** [Agizin/Algorithm-Visualization#11]
Overall, I like Oliver's picture idea and think that is a good direction to move forward in. However, I think there needs to be a stronger distinction between the function of determining the layout of the picture and the actual "drawing" of it to an SVG document. I think we should have two classes:

- A picture class which holds the contents of the data structure and determines the layout of the drawing. Each type of data structure will subclass this. We can lay out a picture by recursively laying out each "subpicture" with relative coordinates. We then lay out the structure of the current picture with a bounding box for each subpicture. Lastly, we adjust the coordinates of each subpicture so that they fit within our bounding box.
- A visualizer class which interfaces with svgwrite and performs the actual drawing based on the layout specified by the picture. This drawing function should only need to be called once per svg file, after the layout of the pictures has been determined. All functions that are performed "at the time of drawing" would be functions of the visualizer.

Is this reasonable? If so, we should start implementation.

**Comment by Byvire** [comment]
Data structures shouldn't subclass Picture.  I'd rather have DataStructures model DataStructures, and let DataStructures be _used_ to make Pictures.  In my head, those concepts (data structure and picture) are cleanly separated, so I'd like to keep them cleanly separated in code as well.  Plus, we don't configure the drawing in any way (including choosing the drawing algorithm) until after the JSON is loaded, whereas a Picture should be put together by a drawing algorithm.
I'm not sure if it matters whether drawing algorithms are implemented as functions that return `Picture`s or as `__init__` methods on subclasses of `Picture` (of `InternalPicture` in particular, whereas `LeafPicture` would have its own subclasses).  I wanted to see if we could get away with the functional approach, and so I wrote the thing below to test it out.  I concluded that drawing algorithms could at least benefit from using subclasses of InternalPicture.
```python
VERTICAL_SPACE = 3  # three times the radius of a node, I guess
HORIZ_SPACE = 1
def draw_tree(root):
    # Draw a tree with uniformly-sized nodes
    root_pic = make_single_node_picture(root)  # Here I'm just waving my hands
    sub_pics = [draw_tree(c) for c in root.children]
    # Note that if children is empty, we'll be okay because we only touch children in loops
    # Make sure the tree nodes are all the same size
    # Absolute scale doesn't matter, so we can just scale each picture so its 
    # nodes (which are of uniform size) are the same size as our root node.
    for pic in sub_pics:
        pic.scale(root_pic.node_radius / pic.node_radius)  # scale the picture in place
    full_pic = InternalPicture(
        width=max(HORIZ_SPACE * root_pic.node_radius * (len(sub_pics) - 1)
                   + sum(pic.width for pic in sub_pics)),
                   root_pic.width),
        height=(VERT_SPACE * root_pic.node_radius
                 + root_pic.height
                 + max((pic.height for pic in sub_pics), default=0)),
        node_radius=root_pic.node_radius,  # This is silly, and there are a few different ways we could fix it
    )
    # I'm guessing we'll want the standard coordinates where the origin (0, 0) is the top-left corner.
    # Here I also assume I can place a sub-picture in a larger picture by specifying coordinates for the sub-picture's origin to land.  We may wish for other ways to insert sub-pictures, e.g. by saying "point (x,y) in the sub-picture's coordinate system should align with point (u,v) in the super-picture's coordinate system."
    full_pic.insert(root_pic, x=(full_pic.width - root_picture.width) / 2, y=0)
    full_pic._root_pic = root_pic  # This makes me worry that this code should be a method on a Picture subclass.
    subtree_y = root_pic.height + VERT_SPACE * root_pic.node_radius
    subtree_next_x = 0
    for pic in sub_pics:
        full_pic.insert(pic, x=subtree_next_x, y=subtree_y)
        # Draw an arrow from the root to the root of the subtree.
        # (I say "rigid arrow" because it's a sort of arrow that shouldn't be shuffled around
        # when all the memory-reference arrows get drawn (depending on the Style).  In our
        # default style, this arrow will always be straight.  Ideas for better names are welcome.)
        full_pic.add_fluff(RigidArrow(root_pic, pic._root_pic))
    return full_pic
```
There's a strong argument to be made for incorporating that function into a subclass of `InternalPicture`.  We can add `node_radius` as a method, like:
```python
@property
def node_radius(self):
    return self._root_pic.width / 2
```
Ideally our interface for drawing algorithms is "callable that returns a Picture", which means we can use subclasses or functions as we wish.  The only other thing I'll say is that I'd like to preserve the feeling that you don't have to do any of the really hard work when you write a drawing algorithm.

It took me a while to figure out why you might want the SVG-drawing code not to be part of the Picture class (or subclasses), even though we don't plan to support any other image format.  It's for Style, right?  I like the idea of having a `Style` class/function/module that handles the style-specific parts of the drawing (i.e. what lines go on the SVG).  Specifically, I guess it can take a list of LeafPictures (converted to absolute coordinates) along with visual fluff and turn the whole bundle into a pretty SVG.

**Comment by Agizin** [comment]
Data structures shouldn't subclass Picture but there should be a subclass of Picture for each data structure. That is, we should have a BinaryTreePicture subclass of Picture that knows how to create a layout for a BinaryTree data structure. Having a method in the Picture class for the layout of each data structure would get too cumbersome as we add data structures and it wouldn't be nearly as easy to add support for more data structures in the future.

I want to separate the layout of the picture from the drawing of the picture to SVG because it feels like a natural division to me. The task of laying out the picture should be independent of how we write that to an SVG. And one of the benefits of this is style: the Picture class shouldn't be concerned with the specifics of the SVG visual: what font we use, what color the lines are, etc (assuming that this is what you mean by style). We can have the Visualizer deal with all that. This should also make it easier to switch between style presets and to change aspects of the visualization based on user settings.

Even conceptually, I think that it is important for us to make this distinction. This is semantics but I think it's an important distinction - our "drawing" algorithm shouldn't return a Picture. Drawing is the creation of the SVG. Our layout algorithm should return a Picture that our drawing algorithm can then traverse and create an SVG from. This is how it works in my mind at least.

Why do we need InternalPicture and LeafPicture to subclass Picture? I would think that we would just need an is_leaf flag variable.

**Comment by Byvire** [comment]
You draw an internal picture by drawing its "visual fluff" (arrows and other crap) and by drawing its sub-pictures.
A leaf picture of a certain type is drawn the same way every time, modulo style.  For instance, it might be a picture of a string or number.

**Comment by Byvire** [comment]
Another reason:  pictures of a given type either will always be leaves (e.g. pictures of string, leaf, null, widget, int, float) or will always be internal (e.g. anything that stores some kind of data -- although we may wish to have leaf versions of certain data structures).  Since we agree that it'd be nice to use the visitor pattern for the actual SVG-drawing code -- that is, to keep all the code for each drawing style together (maybe in a `Style` class) instead of putting it in the Picture classes -- the implementation of a LeafNode subclass can just look like
```python
class NullPicture(LeafPicture):
    pass
```
I actually have a few other things to say, but I'll be a minute.

**Comment by Byvire** [comment]
> Data structures shouldn't subclass Picture but there should be a subclass of Picture for each data structure.

We will have at least one subclass of Picture for each data structure.  For some data structures, we will have more than one.  For instance, we might have a rigidly-spaced n-ary tree, with empty spaces indicating missing nodes, and a condensed tree where the nodes on each level are packed with no space between them.

> I want to separate the layout of the picture from the drawing of the picture to SVG because it feels like a natural division to me. The task of laying out the picture should be independent of how we write that to an SVG. And one of the benefits of this is style: the Picture class shouldn't be concerned with the specifics of the SVG visual: what font we use, what color the lines are, etc (assuming that this is what you mean by style). We can have the Visualizer deal with all that.

This plan has been bugging me now that I've thought about it more.
We want to lay out the picture without considering stylistic decisions like the font size/shape.  But the shape of our String/Int/Float Leaf pictures, and maybe others (any _variable-size_ leaf pictures), will depend intimately on the font shape (and maybe other stylistic decisions, like the shape of the space where the string is supposed to fit).  There are ways for us to work around that without violating our strict separation of Picture and SVG, but all the easy ones I can think of are unsatisfying because they make the drawing uglier and involve hard-coding a font:
- Lay the picture out based on whatever font shape we hard-code.  When making an SVG, simply let strings spill over the edges of their containers.  (Ugly, but acceptable for a prototype.)
- Use monospace fonts exclusively, and guess (hard-code) the width-height ratio of the characters.  If we find out we were wrong, scale the whole picture by `actual_height / guessed_height` vertically and by `actual_width / guessed_width` horizontally.  (This way we lose all control over aspect ratio.  Also, the user will find that if two graphics use different fonts, they won't look similar at all.  And it's needlessly ugly.)
- Of course we could also fix one Official Font of AlgViz and forbid all other fonts, but that's a deal-breaker for usability.  (Exclusively Comic Sans.)

To me, it seems we should consider relaxing our Picture-and-SVG separation.  We won't get rid of the separation, but we should use the Style to produce our Leaf Pictures.

In a sense, the separation between Picture and Style would be a matter of code-reuse:  Everything that doesn't depend on the Style gets implemented in the Picture (and never re-implemented as a result).  Whether to make them the same class is up to you, if you're implementing them.  It might still be nice to make them separate classes (Picture _has_ a Style) so it's easier to use the same Picture subclass with different Styles.

> Visualizer

We should forbid the word "Visualizer".  We only name something "Visualizer" if we don't know what to call it.  Use a more specific name, like "StyleEngine" or "SVGEngine".

**Comment by Agizin** [comment]
The leaf/internal picture architecture makes sense. I had a different picture in my head but I like yours better.

I'm fine with a relaxed separation between picture and style. What I think we need is a relatively hard separation between the act of laying out the coordinates for a picture and the actual drawing of that to an SVG.

Visualizer is a good word and I know exactly what it means when I use it - it's the portion of our code responsible for creating the SVG visual of the data structure. I can go with a different name if it really gets under your skin though, SVGEngine might be fine. To be honest, I think that Picture is misleadingly named - because it isn't meant to be an abstraction of our SVG file (i.e. the only picture we interact with).

I can begin writing an implementation for this but I have a lot going on, so I'm not sure when it'll be finished.

**Comment by Agizin** [comment]
By the way, I found a cool module, https://github.com/btel/svg_utils, that can compose svg files into one document. I need to test it a bit more but this should make drawing to an SVG much easier: we just need to create a new svg for each subpicture and then merge them into one.

**Comment by Agizin** [comment]
I created a branch called picture and I added what I have done there. I still need to do a lot of debugging, it's not working yet for more than a single node with a centered subpicture. However, whether or not it's actually working, it should all be there, except for arrow drawing which we'll need to tackle next. I'll try to finish it up soon, I doubt it'll take much longer. After that, it needs to be reworked a bit to be more visually appealing and we need to implement the arrow drawing.
